qiskit/algorithms/linear_solvers/hhl.py

```diff
@@ -0,0 +1,373 @@
+            r = 2 * constant / a + np.sqrt(np.abs(1 - (2 * constant / a) ** 2))
+            degree = min(nb, int(np.log(1 + (16.23 * np.sqrt(np.log(r) ** 2 + (np.pi / 2) ** 2) *
+                                             kappa * (2 * kappa - self._epsilon_r)) /
+                                        self._epsilon_r)))
+            num_intervals = int(np.ceil(np.log((num_values - 1) / a) / np.log(5)))
+
+            # Calculate breakpoints and polynomials
+            breakpoints = []
+            for i in range(0, num_intervals):
+                # Add the breakpoint to the list
+                breakpoints.append(a * (5 ** i))
+
+                # Define the right breakpoint of the interval
+                if i == num_intervals - 1:
+                    breakpoints.append(num_values - 1)
+
+            reciprocal_circuit = PiecewiseChebyshev(lambda x: np.arcsin(constant / x), degree,
+                                                    breakpoints, nl)
+            na = max(matrix_circuit.num_ancillas, reciprocal_circuit.num_ancillas)
+
+        # Initialise the quantum registers
+        qb = QuantumRegister(nb)  # right hand side and solution
+        ql = QuantumRegister(nl)  # eigenvalue evaluation qubits
+        if na > 0:
+            qa = AncillaRegister(na)  # ancilla qubits
+        qf = QuantumRegister(nf)  # flag qubits
+
+        if na > 0:
+            qc = QuantumCircuit(qb, ql, qa, qf)
+        else:
+            qc = QuantumCircuit(qb, ql, qf)
+
+        # State preparation
+        qc.append(vector_circuit, qb[:])
+        # QPE
+        phase_estimation = PhaseEstimation(nl, matrix_circuit)
+        if na > 0:
+            qc.append(phase_estimation, ql[:] + qb[:] + qa[:matrix_circuit.num_ancillas])
+        else:
+            qc.append(phase_estimation, ql[:] + qb[:])
+        # Conditioned rotation
+        if self._exact_reciprocal:
+            qc.append(reciprocal_circuit, ql[::-1] + [qf[0]])
+        else:
+            qc.append(reciprocal_circuit.to_instruction(), ql[:] + [qf[0]] +
+                      qa[:reciprocal_circuit.num_ancillas])
+        # QPE inverse
+        if na > 0:
+            qc.append(phase_estimation.inverse(), ql[:] + qb[:] +
+                      qa[:matrix_circuit.num_ancillas])
+        else:
+            qc.append(phase_estimation.inverse(), ql[:] + qb[:])
+        return qc
+
+    def solve(self, matrix: Union[np.ndarray, QuantumCircuit],
+              vector: Union[np.ndarray, QuantumCircuit],
+              observable: Optional[Union[LinearSystemObservable, BaseOperator,
+                                         List[LinearSystemObservable], List[BaseOperator]]] = None,
+              observable_circuit: Optional[Union[QuantumCircuit, List[QuantumCircuit]]] = None,
+              post_processing: Optional[Callable[[Union[float, List[float]]],
+                                                 Union[float, List[float]]]] = None) \
+            -> LinearSolverResult:
```

Comment:
Like the EstimationProblem that was done for AE does it make sense to have a LinearProblem here?

Reply:
We already discussed about this and in the end decided against the LinearProblem.

Reply:
Ok,

Reply:
Ok